Add database

lyrapanged opened the pull request.

lyrapanged committed:
initial commit

lyrapanged committed:
12.001

lyrapanged committed:
12.002

lyrapanged committed:
12.003

lyrapanged committed:
12.004

lyrapanged committed:
12.005

lyrapanged committed:
12.006

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/film/Film.java:
```diff
@@ -26,15 +28,24 @@ public class Film {
+    private Mpa mpa;
+    private Set<Genre> genres = new HashSet<>();
 
     public void removeLike(Integer idUser) {
```
Метода по работе с лайками быть больше не должно, так как теперь они хранятся в таблице БД)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/film/Film.java:
```diff
@@ -26,15 +28,24 @@ public class Film {
     @Max(5000)
     private int duration;
     private Set<Integer> filmLikes = new HashSet<>();
```
Поле со списком лайкнувших фильм пользователей следует так же убрать

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/film/Genre.java:
```diff
@@ -0,0 +1,16 @@
+@NoArgsConstructor
+public class Genre {
+    private Integer id;
+    @NotNull
```
Более корректно было бы использовать @NotBlank, так как `@NotNull` допустит пустую строку, либо же ту, что состоит из пробелов)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/film/Mpa.java:
```diff
@@ -0,0 +1,14 @@
+@AllArgsConstructor
+public class Mpa {
+    private Integer id;
+    @NotNull
```
Более корректно было бы использовать @NotBlank, так как `@NotNull` допустит пустую строку, либо же ту, что состоит из пробелов)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/user/User.java:
```diff
@@ -36,7 +38,12 @@ public void setFriends(Integer id) {
         friends.add(id);
     }
 
```
Коллекцию друзей так же следует убрать, так же как и метод, для работы с ней)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/film/MpaService.java:
```diff
@@ -0,0 +1,28 @@
+    private final MpaStorage mpaStorage;
+
+    public Collection<Mpa> getAllMpa() {
+        return mpaStorage.getAllMpa().stream()
```
Более корректно сортировку выполнить на уровне БД, так как за счет встроенных оптимизаций это сделать будет в разы дешевле в плане ресурсов и времени)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/film/GenreService.java:
```diff
@@ -0,0 +1,38 @@
+    private final GenreStorage genreStorage;
+
+    public Collection<Genre> getGenres() {
+        return genreStorage.getGenres().stream()
```
Более корректно сортировку выполнить на уровне БД, так как за счет встроенных оптимизаций это сделать будет в разы дешевле в плане ресурсов и времени)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -35,38 +45,30 @@ public void addFriend(Integer firstId, Integer secondId) {
+        return new ArrayList<>(friendStorage.getFriends(id));
     }
 
     public List<User> commonFriends(Integer firstId, Integer secondId) {
```
Получение списка общих друзей следует произвести с помощью запроса к БД, а не с помощью проверок в коде, так как может потребоваться сравнить намного большее количество данных, из-за чего ресурсов может просто не хватить, так как сначала их нужно получить, десериализовать, потом отфильтровать, а за счет встроенных оптимизаций намного быстрее это сделать на уровне БД, а затем получить уже валидные данные)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -1,13 +1,13 @@
-@Component
+@Repository
 @Slf4j
 public class InMemoryFilmStorage implements FilmStorage {
```
Более не будет использоваться, можно исключить)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -1,13 +1,13 @@
 import java.util.*;
 
-@Component
+@Repository
```
Более не будет использоваться, можно исключить)

TyutterinYakov commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,62 @@
+    PRIMARY KEY (id_film, id_user)
+);
+
+CREATE TABLE IF NOT EXISTS friends
```
Первичный ключ в связующей таблице должен быть составным, либо же следует ввести дополнительный столбец)

TyutterinYakov commented on src/test/java/ru/yandex/practicum/filmorate/FilmorateApplicationTests.java:
```diff
@@ -1,13 +1,18 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
 import org.springframework.boot.test.context.SpringBootTest;
 
```
Отсутствуют тесты, требуемые в ТЗ

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/UserDbStorage.java:
```diff
@@ -0,0 +1,85 @@
+    @Override
+    public List<User> getAllUsers() {
+        String sql = "SELECT * FROM users";
+        return jdbcTemplate.query(sql, (rs, rowNum) -> new User(
```
Для этой цели следует объявить отдельный метод-маппер, чтобы не дублировать код в каждом методе, где требуется получение объекта из RowSet)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/FriendStorage.java:
```diff
@@ -0,0 +1,63 @@
+        userStorage.getUser(userId).orElseThrow(() -> new NotFoundException("User id does not exist"));
+        String sql = "SELECT ID_FRIEND, email, login, name, birthday FROM friends" +
+                " INNER JOIN users ON friends.ID_FRIEND = users.ID_USER WHERE friends.ID_USER = ?";
+        return jdbcTemplate.query(sql, (rs, rowNum) -> new User(
```
Для этой цели следует объявить отдельный метод-маппер, чтобы не дублировать код в каждом методе, где требуется получение объекта из RowSet)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/MpaStorage.java:
```diff
@@ -0,0 +1,41 @@
+        Mpa mpa;
+        SqlRowSet mpaRows = jdbcTemplate.queryForRowSet("SELECT * FROM MPA_RATING WHERE ID_RATING = ?", mpaId);
+        if (mpaRows.first()) {
+            mpa = new Mpa(
```
Для этой цели следует объявить отдельный метод-маппер, чтобы не дублировать код в каждом методе, где требуется получение объекта из RowSet)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/LikeStorage.java:
```diff
@@ -0,0 +1,52 @@
+                "GROUP BY FILMS.ID_FILM ORDER BY COUNT(FILM_LIKES.ID_USER) DESC LIMIT ?";
+
+        return jdbcTemplate.query(getPopularQuery, (rs, rowNum) -> new Film(
+                        rs.getInt("id_film"),
```
Для этой цели следует объявить отдельный метод-маппер, чтобы не дублировать код в каждом методе, где требуется получение объекта из RowSet)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/GenreStorage.java:
```diff
@@ -0,0 +1,68 @@
+
+    public List<Genre> getGenres() {
+        String sql = "SELECT * FROM GENRES";
+        return jdbcTemplate.query(sql, (rs, rowNum) -> new Genre(
```
Для этой цели следует объявить отдельный метод-маппер, чтобы не дублировать код в каждом методе, где требуется получение объекта из RowSet)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/GenreStorage.java:
```diff
@@ -0,0 +1,68 @@
+
+    public void add(Film film) {
+        if (film.getGenres() != null) {
+            for (Genre genre : film.getGenres()) {
```
В данном случае следует использовать `jdbcTemplate.batchUpdate(`, чтобы не плодить лишние запросы к БД на каждой итерации цикла, а сразу же собрать один пакет и отправить его целиком)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,121 @@
+@RequiredArgsConstructor
+public class FilmDbStorage implements FilmStorage {
+    private final JdbcTemplate jdbcTemplate;
+    private final MpaService mpaService;
```
Уровень хранилища не может зависеть от другого хранилища) Отличным вариантом структуры является та, у которой каждый слой зависит только от слоя более низкого уровня, но допускается зависимость сервиса от другого) Так же уровень нижнего уровня не может зависеть от слоя верхнего уровня

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/LikeStorage.java:
```diff
@@ -0,0 +1,52 @@
+@RequiredArgsConstructor
+public class LikeStorage {
+    private final JdbcTemplate jdbcTemplate;
+    private final MpaService mpaService;//f
```
Уровень хранилища не может зависеть от другого хранилища) Отличным вариантом структуры является та, у которой каждый слой зависит только от слоя более низкого уровня, но допускается зависимость сервиса от другого) Так же уровень нижнего уровня не может зависеть от слоя верхнего уровня

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/FriendStorage.java:
```diff
@@ -0,0 +1,63 @@
+@Repository
+public class FriendStorage {
+    private final JdbcTemplate jdbcTemplate;
+    private final UserStorage userStorage;
```
Уровень хранилища не может зависеть от другого хранилища) Отличным вариантом структуры является та, у которой каждый слой зависит только от слоя более низкого уровня, но допускается зависимость сервиса от другого) Так же уровень нижнего уровня не может зависеть от слоя верхнего уровня

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,121 @@
+    public List<Film> getFilms() {
+        String sql = "SELECT * FROM FILMS";
+        return jdbcTemplate.query(sql, (rs, rowNum) -> new Film(
+                rs.getInt("id_film"),
```
Для этой цели следует объявить отдельный метод-маппер, чтобы не дублировать код в каждом методе, где требуется получение объекта из RowSet)

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше на данном этапе :) Хорошего вечера! Если возникнут вопросы - пиши в слак, буду рад помочь! :) Кстати, в схеме на скрине в таблице films у Id_rating должен быть другой тип)

lyrapanged committed:
12.006 -right

lyrapanged committed:
12.006 (User)remove Set friend

lyrapanged committed:
12.006 without test

lyrapanged committed:
12.1002 reformat Services,add Test

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -38,24 +48,21 @@ public List<Film> getFilms() {
-        film.setFilmLikes(user.getId());
-        film.setLikesCounter(film.getFilmLikes().size());
+    public void addLike(Integer filmId, Integer userId) {
+        filmStorage.getFilm(filmId).orElseThrow(() -> new NotFoundException("Film or film doesn't exist"));
```
Можно было бы вызывать метод нашего сервиса - getFilm, чтобы не дублировать код с выбросом исключения)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -38,24 +48,21 @@ public List<Film> getFilms() {
-        film.setLikesCounter(film.getFilmLikes().size());
+    public void addLike(Integer filmId, Integer userId) {
+        filmStorage.getFilm(filmId).orElseThrow(() -> new NotFoundException("Film or film doesn't exist"));
+        userStorage.getUser(userId).orElseThrow(() -> new NotFoundException("User or film doesn't exist"));
```
Можно было бы внедрить сервис пользователей, чтобы вызывать метод сервиса - getUser, чтобы не дублировать код выброса исключения) Либо же можно было бы объявить приватный метод в сервисе FilmService, чтобы вызывать его в тех местах, где требуется проверить наличие пользователя

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -38,24 +48,21 @@ public List<Film> getFilms() {
-        film.removeLike(user.getId());
-        film.setLikesCounter(film.getFilmLikes().size());
+    public void removeLike(Integer filmId, Integer userId) {
+        userStorage.getUser(userId).orElseThrow(() -> new NotFoundException("User or film doesn't exist"));
```
Аналогично и здесь)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    public List<Film> getFilms() {
+        String sql = "SELECT * FROM FILMS";
```
Получение MPA следует произвести одним SQL запросом)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,108 @@
+        if (filmId == null) {
+            throw new NotFoundException("Id doesn't exist");
+        }
+        String sql = "SELECT * FROM films WHERE ID_FILM = ?";
```
Получение MPA следует произвести одним SQL запросом)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,108 @@
+        String login = rs.getString("description");
+        LocalDate releaseDate = rs.getDate("release_Date").toLocalDate();
+        Integer duration = rs.getInt("duration");
+        String sqlNameMpa = "SELECT * FROM MPA_RATING WHERE ID_RATING IN (" +
```
Логику по получению MPA следует исключить, так как получение MPA будет выполняться в одном SQL запросе и его извлечение можно будет произвести из того же ResultSet'a)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmDbStorage.java:
```diff
@@ -0,0 +1,108 @@
+                        (rsMpa, rowNum) -> rsMpa.getString("name"), idFilm).stream()
+                .findAny().orElseThrow(() -> new NotFoundException("Bad name"));
+        Mpa mpa = new Mpa(rs.getInt("id_rating"), nameMpa);
+        String sqlGenres = "SELECT * FROM GENRES WHERE ID_GENRE IN " +
```
Жанры следует присоединить к фильмам на уровне сервиса фильмов, обратившись к DAO жанров для их получения)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/LikeStorage.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    public List<Film> getPopularFilms(Integer count) {
+        String getPopularQuery = "SELECT FILMS.ID_FILM, NAME, DESCRIPTION, RELEASE_DATE, DURATION, ID_RATING " +
```
Получение MPA следует произвести одним SQL запросом)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/LikeStorage.java:
```diff
@@ -0,0 +1,71 @@
+        String login = rs.getString("description");
+        LocalDate releaseDate = rs.getDate("release_Date").toLocalDate();
+        Integer duration = rs.getInt("duration");
+        String sqlNameMpa = "SELECT * FROM MPA_RATING WHERE ID_RATING IN (" +
```
Логику по получению MPA следует исключить, так как получение MPA будет выполняться в одном SQL запросе и его извлечение можно будет произвести из того же ResultSet'a)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/LikeStorage.java:
```diff
@@ -0,0 +1,71 @@
+                        (rsMpa, rowNum) -> rsMpa.getString("name"), idFilm).stream()
+                .findAny().orElseThrow(() -> new NotFoundException("Bad name"));
+        Mpa mpa = new Mpa(rs.getInt("id_rating"), nameMpa);
+        String sqlGenres = "SELECT * FROM GENRES WHERE ID_GENRE IN " +
```
Жанры следует присоединить к фильмам на уровне сервиса фильмов, обратившись к DAO жанров для их получения)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/FriendStorage.java:
```diff
@@ -0,0 +1,89 @@
+@Repository
+public class FriendStorage {
+    private final JdbcTemplate jdbcTemplate;
+    private final UserStorage userStorage;
```
Не исправлено - "Уровень хранилища не может зависеть от другого хранилища)"
Всех их взаимодействие должно осуществляться на уровне сервиса)

TyutterinYakov reviewed (commented):
Добрый вечер! Отметил еще некоторые недочеты :) Хорошего вечера!

lyrapanged committed:
12.200001

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/ErrorHandler.java:
```diff
@@ -39,10 +39,10 @@ public ResponseEntity<String> ex(final MethodArgumentNotValidException e) {
-        log.error("Unhandled exception");
-        return new ResponseEntity<>(e.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR);
-    }
+//    @ExceptionHandler
```
Обработку следует оставить, так как необоходимо ловить и логгировать так же все необработанные исключение другими хендлерами

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/FriendStorage.java:
```diff
@@ -0,0 +1,77 @@
+
+@RequiredArgsConstructor
+@Repository
+public class FriendStorage {
```
У этого DAO так же должен быть интерфейс, чтобы мы могли не привязываться к конкретной реализации хранилища, а работать лишь с предоставленным API

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/MpaStorage.java:
```diff
@@ -0,0 +1,38 @@
+@Repository
+@RequiredArgsConstructor
+@Validated
+public class MpaStorage {
```
У этого DAO так же должен быть интерфейс, чтобы мы могли не привязываться к конкретной реализации хранилища, а работать лишь с предоставленным API

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/MpaStorage.java:
```diff
@@ -0,0 +1,38 @@
+
+@Repository
+@RequiredArgsConstructor
+@Validated
```
Аннотация `@Validated` здесь не требуется, так как отсутствуют аннотации валидации для значений аргументов методов)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/GenreStorage.java:
```diff
@@ -0,0 +1,75 @@
+
+@Repository
+
+public class GenreStorage {
```
У этого DAO так же должен быть интерфейс, чтобы мы могли не привязываться к конкретной реализации хранилища, а работать лишь с предоставленным API

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/GenreStorage.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.List;
+
+@Repository
+
```
Лишняя пустая строка

TyutterinYakov reviewed (commented):
Добрый день! Отметил оставшиеся недочеты :) Жду на финальное ревью) Хорошего дня!

lyrapanged committed:
12.300001 changed filmorate/storage to filmorate/dao
changed filmorate/storage/user to filmorate/dao/userDao
changed filmorate/storage/film to filmorate/dao/filmDao

lyrapanged commented:
Добрый день. Если я в коде допускаю ошибку, то при прохождении тестов, обработка исключений типа Throwable выдает только сообщение из логера и для того чтобы увидеть детальное описание ошибки я отключаю этот обработчик. Возникает возможность забыть включить его обратно. Есть ли более лучшее решения для просмотра полного описания ошибки?

TyutterinYakov commented:
Более подробно с методами Throwable можно ознакомиться тут https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/lang/Throwable.html
Каждый описан достаточно подробно и думаю, что ты найдешь ответ на свой вопрос) Можешь попробовать воспользоваться всеми и посмотреть, что выводит каждый из них)

TyutterinYakov reviewed (approved):
Добрый день! Поздравляю с успешной сдачей спринта! Сил и терпения на этом сложном, но интересном пути! Хорошего дня и отличного настроения🌅
